### ISSUE TITLE
Explicit annotation ignored in type argument inference

### ISSUE THREAD

**Issue by smillst** [typetools/checker-framework#8284]
```java
// FALSE POSITIVE.  Reduced from Beam JavaBeanSchema.java:176, JavaFieldSchema.java:122,
// AutoValueSchema.java:93 (all three report the identical error).
//
//   error: [argument] incompatible argument for parameter typeDescriptor of getGetters.
//     found   : TypeDescriptor<T extends @Nullable Object>
//     required: TypeDescriptor<@NonNull T extends @NonNull Object>
//
// `@NonNull` is written on a *use* of T in the return type.  When the call has a return/assignment
// target, that `@NonNull` contaminates the inferred type argument itself, so the *parameter* type
// becomes TypeDescriptor<@NonNull T> and no longer accepts the argument TypeDescriptor<T>.
//
// Controls below all pass, which is what makes this a false positive:
//   - spelling out the type argument explicitly
//   - dropping the `@NonNull` from the return type
//   - removing the return-type target context
package beamerrors;

import java.util.List;
import org.checkerframework.checker.nullness.qual.NonNull;

public class NonNullTypeVarInReturnType {
  static class TypeDescriptor<T> {}

  static class Getter<K, V> {}

  static <T> List<Getter<@NonNull T, Object>> getGetters(TypeDescriptor<T> typeDescriptor) {
    throw new RuntimeException();
  }

  // Errors.
  static <T> List<Getter<@NonNull T, Object>> caller(TypeDescriptor<T> targetTypeDescriptor) {
    return getGetters(targetTypeDescriptor);
  }

  // Control: explicit type argument.  No error.
  static <T> List<Getter<@NonNull T, Object>> explicitTypeArgument(TypeDescriptor<T> td) {
    return NonNullTypeVarInReturnType.<T>getGetters(td);
  }

  // Control: no @NonNull in the return type.  No error.
  static <T> List<Getter<T, Object>> plainGetters(TypeDescriptor<T> td) {
    throw new RuntimeException();
  }

  static <T> List<Getter<T, Object>> noAnnotation(TypeDescriptor<T> td) {
    return plainGetters(td);
  }

  // Control: no target type.  No error.
  static <T> void noTargetType(TypeDescriptor<T> td) {
    getGetters(td);
  }
}

```
